scripts: extract_dts_includes: Fix extract_cells for a list

galak opened the pull request:
If we had something like:

gpios = <&gpiod 13 GPIO_INT_ACTIVE_LOW
         &gpiod 14 GPIO_INT_ACTIVE_LOW>;

The script blows up in extract_cells.  While extract_cells attempted to
handle more than a single item in such a list, it didn't manipulate the
prop list for the recursive calls properly.

Since we pop off items from the prop list as we use them we can easily
pass the prop list to the rescurive call to fix things.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

agross-korg reviewed (approved):
this looks reasonable

nashif commented:
recheck

erwango commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -228,16 +228,11 @@ def extract_controller(node_address, y_key, prefix, defs, def_label):
 
 
-def extract_cells(node_address, yaml, y_key, names, index, prefix, defs,
+def extract_cells(node_address, props, yaml, y_key, names, index, prefix, defs,
```
y_key is not needed anymore, I think

erwango reviewed (dismissed):
I had a look on what is generated when using this, and there are a couple things to be aware/fix.

I used following dummy example:
```
	leds {
		compatible = "gpio-leds";
		green_led_2: led@0 {
			gpios = <&gpioa 5 GPIO_INT_ACTIVE_HIGH
				 &gpiob 4 GPIO_INT_ACTIVE_HIGH>;
			label = "User LD2";
		};
	};
	aliases {
		led0 = &green_led_2;
	};
```

```
/* led@0 */
#define GPIO_LEDS_0_GPIO_CONTROLLER	"GPIOA"
#define GPIO_LEDS_0_GPIO_CONTROLLER_0	"GPIOA"
#define GPIO_LEDS_0_GPIO_CONTROLLER_1	"GPIOB"
#define GPIO_LEDS_0_GPIO_FLAGS_0	4
#define GPIO_LEDS_0_GPIO_FLAGS_1	4
#define GPIO_LEDS_0_GPIO_PIN_0		5
#define GPIO_LEDS_0_GPIO_PIN_1		4
#define GPIO_LEDS_0_LABEL		"User LD2"
#define GPIO_LEDS_0_GPIO_FLAGS		GPIO_LEDS_0_GPIO_FLAGS_0
#define GPIO_LEDS_0_GPIO_PIN		GPIO_LEDS_0_GPIO_PIN_0
#define LED0_GPIO_CONTROLLER_0		GPIO_LEDS_0_GPIO_CONTROLLER_0
#define LED0_GPIO_CONTROLLER_1		GPIO_LEDS_0_GPIO_CONTROLLER_1
#define LED0_GPIO_FLAGS			GPIO_LEDS_0_GPIO_FLAGS
#define LED0_GPIO_PIN			GPIO_LEDS_0_GPIO_PIN
#define LED0_LABEL			GPIO_LEDS_0_LABEL
```

So one should know that `GPIO_LEDS_0_GPIO_CONTROLLER, GPIO_LEDS_0_GPIO_FLAGS and GPIO_LEDS_0_GPIO_PIN` will refer to first item. Not sure there is something to fix though, as _0 and _1 are correctly generated. 

What I think should be fixed is that `LED0_GPIO_FLAGS and LED0_GPIO_PIN` are not generated for second item (while we get the _0 and _1 versions for _CONTROLLER)

galak commented:
@erwango do you mind taking this over and making the fixes you mentioned.

galak commented on scripts/dts/extract_dts_includes.py:
```diff
@@ -443,7 +438,11 @@ def extract_property(node_compat, yaml, node_address, y_key, y_val, names,
         extract_controller(node_address, y_key, prefix, defs, def_label)
-        extract_cells(node_address, yaml, y_key,
+        try:
+            props = list(reduced[node_address]['props'].get(y_key))
```
I was being a bit lazy in my approach, was wondering if keeping extract_cells and add a second function that extract_cells calls.

erwango commented:
> @erwango do you mind taking this over and making the fixes you mentioned.

Sure, I could do it, is there a real life example to work on?

galak commented:
> Sure, I could do it, is there a real life example to work on?

The case you have is pretty much similar to my real life one, the only difference is its not for led.

erwango commented:
> The case you have is pretty much similar to my real life one, the only difference is its not for led.

One point to check: do we want node to be a gpio-device?
For instance (which is a bit different case, but I think we should start from that):
```
&spi3 {
	status = "ok";

	spbtle-rf@0 {
		compatible = "st,spbtle-rf";
		reg = <0>;
		spi-max-frequency = <8000000>;
		label = "SPBTLE-RF";
		cs-gpios = <&gpiod 13 GPIO_INT_ACTIVE_LOW>;
		reset-gpios = <&gpioa 8 GPIO_INT_ACTIVE_LOW>;
		irq-gpios = <&gpioe 6 GPIO_INT_ACTIVE_LOW>;
		};
};
```

This node is spi-device. We can add cs-gpios, reset-gpios and irq-gpios in its yaml properties,
and with minor script modification (take property name into account in define label generation), we can get things generated.
For instance (not fully working yet, will be easier tomorrow  with fresh mind)
```
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_BASE_ADDRESS_0		0x0
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_BUS_NAME			"SPI_3"
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_CS_GPIOS_CONTROLLER		"GPIOD"
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_GPIO_FLAGS_0			0
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_GPIO_PIN_0			6
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_IRQ_GPIOS_CONTROLLER		"GPIOE"
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_LABEL				"SPBTLE-RF"
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_RESET_GPIOS_CONTROLLER		"GPIOA"
#define ST_STM32_SPI_FIFO_40003C00_ST_SPBTLE_RF_0_SPI_MAX_FREQUENCY		8000000
```

Would you be happy with that (ie not using gpio-device for this kind of case)?
Then, I'll have a look to multiple pins.

erwango committed:
scripts: extract_dts_includes: rename arguments for easier reading

Some functions were using y_key and y_val as argument.  This was not
quite easy to read.  Rename argument into more informative versions.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@galak:

Unlike mentioned in the previous commit, it might be clearer to handle things separately.
So my proposal is the following.
If we have, in led0, something like:
```
gpios = <&gpiod 13 GPIO_INT_ACTIVE_LOW
&gpiod 14 GPIO_INT_ACTIVE_LOW>;
```
generate something like:
```
/* led@0 */
#define GPIO_LEDS_0_GPIO_CONTROLLER_0	"GPIOA"
#define GPIO_LEDS_0_GPIO_CONTROLLER_1	"GPIOB"
#define GPIO_LEDS_0_GPIO_FLAGS_0	4
#define GPIO_LEDS_0_GPIO_FLAGS_1	4
#define GPIO_LEDS_0_GPIO_PIN_0		5
#define GPIO_LEDS_0_GPIO_PIN_1		4
#define GPIO_LEDS_0_LABEL		"User LD2"
#define LED0_GPIO_CONTROLLER_0		GPIO_LEDS_0_GPIO_CONTROLLER_0
#define LED0_GPIO_CONTROLLER_1		GPIO_LEDS_0_GPIO_CONTROLLER_1
#define LED0_GPIO_FLAGS_0			GPIO_LEDS_0_GPIO_FLAGS_0
#define LED0_GPIO_FLAGS_1			GPIO_LEDS_0_GPIO_FLAGS_1
#define LED0_GPIO_PIN_0			GPIO_LEDS_0_GPIO_PIN_0
#define LED0_GPIO_PIN_1			GPIO_LEDS_0_GPIO_PIN_1
#define LED0_LABEL			GPIO_LEDS_0_LABEL
```
Which means:
-only the duplicate objects have an index (gpio_controller, pins and flags in this case)
-aliases indexes are generated on the same basis (ie no more alias on the item _0)

According to @agross-korg aliases to _0 index were implemented for a purpose which isn't in use currently, so this shouldn't cause any problem.
Are you fine with that?

galak committed:
scripts: extract_dts_includes: Fix extract_cells for a list

If we had something like:

gpios = <&gpiod 13 GPIO_INT_ACTIVE_LOW
         &gpiod 14 GPIO_INT_ACTIVE_LOW>;

The script blows up in extract_cells.  While extract_cells attempted to
handle more than a single item in such a list, it didn't manipulate the
prop list for the recursive calls properly.

Since we pop off items from the prop list as we use them we can easily
pass the prop list to the rescurive call to fix things.

Besides, if alias is defined for the node, indexed aliases defines are
generated:
 #define GPIO_LEDS_0_GPIO_FLAGS_0 4
 #define GPIO_LEDS_0_GPIO_FLAGS_1 0
 #define GPIO_LEDS_0_GPIO_PIN_0   5
 #define GPIO_LEDS_0_GPIO_PIN_1   6
 #define LED1_GPIO_FLAGS_0 GPIO_LEDS_0_GPIO_FLAGS_0
 #define LED1_GPIO_FLAGS_1 GPIO_LEDS_0_GPIO_FLAGS_1
 #define LED1_GPIO_PIN_0 GPIO_LEDS_0_GPIO_PIN_0
 #define LED1_GPIO_PIN_1 GPIO_LEDS_0_GPIO_PIN_1

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

erwango commented:
@galak , I've pushed a new version of the PR according to my last comment.

erwango committed:
scripts: extract_dts_includes: Generate'_0' defines only when needed

Indexed defines were systematically generated even when there
was only one element to generate.
So we ended up generated a lot of _0 defines.
Then we needed to generate aliases to these _0 indexed defines,
in order to get useful defines.
For instance:
 #define GPIO_LEDS_0_GPIO_FLAGS_0	4
 #define GPIO_LEDS_0_GPIO_PIN_0		5
 #define GPIO_LEDS_0_GPIO_FLAGS		GPIO_LEDS_0_GPIO_FLAGS_0
 #define GPIO_LEDS_0_GPIO_PIN		GPIO_LEDS_0_GPIO_PIN_0

This commit allows to generate _0 indexed define only if a
property has more than one elements to define.
Aliases generation to _0 indexed defines are also removed.

Note: IRQ are left untouched since this is frequent to handle
multiple IRQs in a driver

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

galak committed:
scripts: extract_dts_includes: Fix extract_controller for a list

extract_controller only handle the first phandle in a cell property.
However we could easily have something like this where the phandles
vary:

gpios = <&gpiod 13 GPIO_INT_ACTIVE_LOW
         &gpioc 14 GPIO_INT_ACTIVE_LOW>;

So we need to walk the property list for each phandle and produce a
define associated with it.

Also, if alias to the node is defined, indexed alias define
is generated for every indexed controller define:
 #define GPIO_LEDS_0_GPIO_CONTROLLER_0	"GPIOA"
 #define GPIO_LEDS_0_GPIO_CONTROLLER_1	"GPIOB"
 #define LED1_GPIO_CONTROLLER_0		GPIO_LEDS_0_GPIO_CONTROLLER_0
 #define LED1_GPIO_CONTROLLER_1		GPIO_LEDS_0_GPIO_CONTROLLER_1

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

erwango committed:
scripts: extract_dts_includes: remove prefix argument

prefix was given as argument in several functions but not used
as property is never defined.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
scripts: extract_includes_dts: Remove usage of cell_string yaml attribute

'cell_string' yaml attribute has been introduced in order to
help enforcement of specific string during defines generation.
This adds complexity in understanding script behavior as a black
box and create additional dependency which is not strictly required.

For node specific generation functions (pinctrl and interrupts),
this could be replaced directly by an hardcoded version
(as everyone used the same 'cell_string' anyway).
For extract_cells functions, string could be replaced by extracted
property name. As a consequence, we're now able to generate defines
for properties refering to these controllers via phandle.
For instance, in following node
 	spbtle-rf@0 {
 		compatible = "st,spbtle-rf";
 		reg = <0>;
		reset-gpios = <&gpioa 8 0>;
 	};
We'll be able to generate:
 #define ST_STM32_SPI_...LE_RF_0_RESET_GPIOS_CONTROLLER	"GPIOA"
 #define ST_STM32_SPI_...PBTLE_RF_0_RESET_GPIOS_FLAGS_0		0
 #define ST_STM32_SPI_..._SPBTLE_RF_0_RESET_GPIOS_PIN_0		8

Only impact for this whole change is for NXP clocks which were the
only ones using 'cell_string' attribute with a value different than
the default one.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
arch: nxp dts bindings: Set IP clock name to matching clock controller

In dts.fixup files for used for NXP SoCs, IP clock name was
bound to SIM clock label. This is correct but would require
manual change if IP clock controller was changed in device tree.
Use reference to IP generated define "_CLOCK_CONTROLLER" to get
this value.

Besides, having all IP related values starting with the same prefix
will be easier for possible macro preprocessing (if required one day).

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

GAnthony reviewed (approved):
LGTM
